### PR TITLE
refactor(typescript): packages/rspack/src/template from js to ts and esm

### PR DESCRIPTION
**Pull request by shulaoda**
## Summary
#7225 

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by shulaoda** [comment]
The file `Template.js` seems to have been copied directly from `webpack`, and some `modules` in `Rspack` have not been adapted, so I have removed some unnecessary code.

File: packages/rspack/src/Template.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+"use strict";
```
Comment (SoonIter) [inline packages/rspack/src/Template.ts:5]:
```suggestion
```

**Comment by SoonIter** [comment]
> The file `Template.js` seems to have been copied directly from `webpack`, and some `modules` in `Rspack` have not been adapted, so I have removed some unnecessary code.

it is acceptable i think, could you add `The following code is modified based on` comments at the top ?

https://github.com/web-infra-dev/rspack/blob/de876babb550365b80cf904917471a58af72e541/packages/rspack/src/ErrorHelpers.ts#L1-L9

**Review by SoonIter — APPROVED** [review]
thanks 🫶🏻

**Pull request merged**